cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -367,12 +379,106 @@ class device_row_comparator {
+      weak_ordering state{weak_ordering::EQUIVALENT};
+      for (int left_dremel_index = l_start, right_dremel_index = r_start, element_index = 0;
+           left_dremel_index < l_end and right_dremel_index < r_end;
+           ++left_dremel_index, ++right_dremel_index) {
+        // First early exit: the definition levels do not match.
+        if (l_def_levels[left_dremel_index] != r_def_levels[right_dremel_index]) {
+          state = (l_def_levels[left_dremel_index] < r_def_levels[right_dremel_index])
+                    ? weak_ordering::LESS
+                    : weak_ordering::GREATER;
+          return cuda::std::pair(state, _depth);
+        }
+
+        // Second early exit: the repetition levels do not match.
+        if (l_rep_levels[left_dremel_index] != r_rep_levels[right_dremel_index]) {
+          state = (l_rep_levels[left_dremel_index] < r_rep_levels[right_dremel_index])
+                    ? weak_ordering::LESS
+                    : weak_ordering::GREATER;
+          return cuda::std::pair(state, _depth);
+        }
+
+        // Third early exit: This case has two branches.
+        // 1) If we are at the maximum definition level, then we actually have
+        //    an underlying element to compare, not just an empty list or a
+        //    null. Therefore, we access the kth element of each list and
+        //    compare the values.
+        // 2) If we are 1 - the maximum definition level and the column is
+        //    nullable, we know that we are looking at a element in the child
+        //    column. In this case we simply skip to the next element.
+        if (l_def_levels[left_dremel_index] == l_max_def_level) {
+          int last_null_depth                    = _depth;
+          cuda::std::tie(state, last_null_depth) = cudf::type_dispatcher<dispatch_void_if_nested>(
+            lcol.type(), comparator, element_index, element_index);
+          if (state != weak_ordering::EQUIVALENT) { return cuda::std::pair(state, _depth); }
+          ++element_index;
+        } else if (lcol.nullable() and l_def_levels[left_dremel_index] == l_max_def_level - 1) {
+          ++element_index;
+        }
+      }
+
+      // If we have reached this stage, we know that definition levels,
+      // repetition levels, and actual elements are identical in both list
+      // columns up to the `min(l_end - l_start, r_end - r_start)` element of
+      // the Dremel encoding. However, two lists can only compare equivalent if
+      // they are of the same length. Otherwise, the shorter of the two is less
+      // than the longer. This final check determines the appropriate resulting
+      // ordering by checking how many total elements each list is composed of.
+      state = (l_end - l_start < r_end - r_start)   ? weak_ordering::LESS
+              : (l_end - l_start > r_end - r_start) ? weak_ordering::GREATER
+                                                    : weak_ordering::EQUIVALENT;
```

Comment:
Maybe reuse this function. https://github.com/rapidsai/cudf/blob/c01a2a41b7d57a1360324270101d9304fbc9515f/cpp/include/cudf/table/row_operators.cuh#L62

```suggestion
      state = cudf::detail::compare_elements(l_end - l_start, r_end - r_start);
```